src/lib/client.js

```diff
@@ -26,18 +30,93 @@ module.exports = class BtpClient {
     this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
   }
 
+  tryToOpenWebSocketClient () {
```

Comment:
These three functions are pretty confusing. `tryToOpenWebSocketClient` calls `ensureUpstream`, which calls `getOpenWebSocketClient`, which calls `tryToOpenWebSocketClient`

Reply:
`tryToOpenWebSocketClient` is pretty straightforward and well-named, I think; it doesn't itself call `ensureUpstream`, but it sets a auto-reconnect handler, which calls `ensureUpstream` from the start if the socket closes. There is a code comment there that clarifies that.

`getOpenWebSocketClient` calls the basic `tryToOpen` function repeatedly, from a timer. What is slightly complicated there is the `tryOnce` function which deals with overlapping attempts. The timer is there, instead of just waiting for the previous attempt to end, because I think it's safer that way (although I have no proof of this).

`ensureUpstream` not only gets the open websocket incarnation, it also connects it in both directions to the wrapper that is passed to `plugin.addSocket`. The reason this is necessary is that if you pass the first incarnation of the raw open WebSocket object to the plugin, and then (due to server downtime) the first incarnation is closed and we need to create a new WebSocket client instance (the second incarnation), then the plugin would from then on be talking to the wrong incarnation. This goes for both `ws.on('message') and for `ws.send`.